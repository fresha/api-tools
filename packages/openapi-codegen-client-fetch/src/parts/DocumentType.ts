import { Nullable } from '@fresha/api-tools-core';
import { addImportDeclaration, addTypeAlias } from '@fresha/code-morph-ts';
import { assert } from '@fresha/openapi-codegen-utils';
import { SyntaxKind } from 'ts-morph';

import { NamedType } from './NamedType';
import { ResourceType } from './ResourceType';
import { determineSchemaName, schemaToType } from './utils';

import type { ActionContext } from '../context';
import type { SchemaModel } from '@fresha/openapi-model/build/3.0.3';

export class DocumentType extends NamedType {
  protected isRequestBody: boolean;
  protected primaryResourceTypes: NamedType[];
  protected primaryDataIsArray: boolean;
  protected includedResourceTypes: NamedType[];

  constructor(context: ActionContext, name: string, schema: SchemaModel, isRequestBody: boolean) {
    super(context, name, schema);
    this.isRequestBody = isRequestBody;
    this.primaryResourceTypes = [];
    this.primaryDataIsArray = false;
    this.includedResourceTypes = [];
  }

  collectData(namedTypes: Map<string, NamedType>): void {
    const primaryDataSchema = this.schema.getProperty('data');
    if (primaryDataSchema) {
      this.determinePrimaryDataTypes(primaryDataSchema, namedTypes);
    }
    const includedSchema = this.schema.getProperty('included');
    if (includedSchema) {
      this.determineIncludedDataTypes(includedSchema, namedTypes);
    }
  }

  private determinePrimaryDataTypes(
    primaryDataSchema: Nullable<SchemaModel>,
    namedTypes: Map<string, NamedType>,
  ): void {
    const primaryDataSchemas: SchemaModel[] = [];

    if (primaryDataSchema?.type === 'array') {
      if (Array.isArray(primaryDataSchema.items)) {
        primaryDataSchemas.push(...primaryDataSchema.items);
      } else if (primaryDataSchema.items) {
        primaryDataSchemas.push(primaryDataSchema.items);
      }
      this.primaryDataIsArray = true;
    } else if (primaryDataSchema) {
      primaryDataSchemas.push(primaryDataSchema);
    }

    assert(primaryDataSchema, `Cannot determine type of primary data`, this.context.operation);

    const primaryDataAlternatives = [];
    for (const primaryDataAlt of primaryDataSchemas) {
      primaryDataAlternatives.push(
        ...(primaryDataAlt.oneOf ?? []),
        ...(primaryDataAlt.anyOf ?? []),
      );
    }
    if (primaryDataAlternatives.length) {
      for (const subschema of primaryDataAlternatives) {
        this.determineSinglePrimaryDataType(subschema, namedTypes);
      }
    } else if (primaryDataSchemas.length) {
      for (const subschema of primaryDataSchemas) {
        this.determineSinglePrimaryDataType(subschema, namedTypes);
      }
    }
  }

  private determineSinglePrimaryDataType(
    primaryDataSchema: SchemaModel,
    namedTypes: Map<string, NamedType>,
  ): void {
    const primaryDataSchemaName = determineSchemaName(primaryDataSchema);

    let existingNamedType = namedTypes.get(primaryDataSchemaName);
    assert(
      !existingNamedType || existingNamedType.schema === primaryDataSchema,
      `Name ${primaryDataSchemaName} have duplicate associated schemas`,
      this.context.operation,
    );

    if (!existingNamedType) {
      existingNamedType = new ResourceType(
        this.context,
        primaryDataSchemaName,
        primaryDataSchema,
        this.isRequestBody,
      );
      namedTypes.set(primaryDataSchemaName, existingNamedType);

      existingNamedType.collectData(namedTypes);
    }

    this.primaryResourceTypes.push(existingNamedType);
  }

  private determineIncludedDataTypes(
    includedSchema: SchemaModel,
    namedTypes: Map<string, NamedType>,
  ): void {
    assert(
      includedSchema.type === 'array',
      `'included' field must be an array schema, but it is an ${String(includedSchema.type)}`,
      this.context.operation,
    );

    const includedItemsSchema = includedSchema.items;
    if (!includedItemsSchema) {
      return;
    }

    const includedItemsSchemas: SchemaModel[] = [];
    if (Array.isArray(includedItemsSchema)) {
      includedItemsSchemas.push(...includedItemsSchema);
    } else {
      includedItemsSchemas.push(includedItemsSchema);
    }

    for (const s of includedItemsSchemas) {
      if (s.type === 'object') {
        this.determineSingleIncludedDataType(s, namedTypes);
      } else {
        // multiple types of resources
        const subshemas = [...(s.oneOf ?? []), ...(s.anyOf ?? [])];

        for (const subschema of subshemas) {
          this.determineSingleIncludedDataType(subschema, namedTypes);
        }
      }
    }
  }

  private determineSingleIncludedDataType(
    schema: SchemaModel,
    namedTypes: Map<string, NamedType>,
  ): void {
    let schemaName: string;
    if (schema.parent === schema.root.components) {
      schemaName = schema.title ?? schemaToType(schema);
    } else {
      schemaName = schemaToType(schema);
    }

    let existingNamedType = namedTypes.get(schemaName);
    assert(
      !existingNamedType || existingNamedType.schema === schema,
      `Name ${schemaName} have duplicate associated schemas`,
      this.context.operation,
    );

    if (!existingNamedType) {
      existingNamedType = new ResourceType(this.context, schemaName, schema, this.isRequestBody);
      namedTypes.set(schemaName, existingNamedType);

      existingNamedType.collectData(namedTypes);
    }

    this.includedResourceTypes.push(existingNamedType);
  }

  generateCode(generatedTypes: Set<string>): void {
    if (!generatedTypes.has(this.name)) {
      generatedTypes.add(this.name);
    } else {
      return;
    }

    for (const resourceType of this.primaryResourceTypes) {
      resourceType.generateCode(generatedTypes);
    }
    for (const includedType of this.includedResourceTypes) {
      includedType.generateCode(generatedTypes);
    }

    const typeBaseName = this.isRequestBody ? 'JSONAPIClientDocument' : 'JSONAPIDataDocument';
    addImportDeclaration(this.context.typesFile, '@fresha/api-tools-core', `t:${typeBaseName}`);
    const typeAlias = addTypeAlias(this.context.typesFile, this.name, typeBaseName, true);

    const typeAliasArgs = [];
    if (this.primaryResourceTypes.length) {
      const typeName = this.primaryResourceTypes.map(t => t.name).join(' | ');
      typeAliasArgs.push(this.primaryDataIsArray ? `(${typeName})[]` : `${typeName}`);
    }
    if (this.includedResourceTypes.length) {
      if (!this.primaryResourceTypes.length) {
        typeAliasArgs.push(this.isRequestBody ? 'JSONAPIClientResource' : 'JSONAPIServerResource');
      }
      const names = this.includedResourceTypes.map(t => t.name).join(' | ');
      typeAliasArgs.push(`${names}`);
    }
    typeAlias
      .getNodeProperty('type')
      .asKindOrThrow(SyntaxKind.TypeReference)
      .addTypeArguments(typeAliasArgs);

    generatedTypes.delete(this.name);
  }
}
