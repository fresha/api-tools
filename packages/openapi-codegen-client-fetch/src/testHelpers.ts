// eslint-disable-next-line import/no-extraneous-dependencies
import { createTSProjectTestContext } from '@fresha/openapi-codegen-test-utils';

import type { ActionContext, Context } from './context';
import type { OpenAPIModel, OperationModel } from '@fresha/openapi-model/build/3.0.3';

export const createTestContext = (openapi: OpenAPIModel): Context => {
  const base = createTSProjectTestContext(openapi);

  return {
    ...base,
    includeDeprecated: false,
    includedTags: new Set<string>(),
    excludedTags: new Set<string>(),
    transformRequest: 'kebab',
    transformResponse: 'camel',
  };
};

export const createActionTestContext = (
  operation: OperationModel,
  fileName: string,
): ActionContext => {
  const base = createTestContext(operation.root);
  const sourceFile = base.project.createSourceFile(fileName, '');

  return {
    ...base,
    sourceFile,
    operation,
  };
};
