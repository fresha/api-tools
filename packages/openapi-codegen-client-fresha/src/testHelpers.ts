// eslint-disable-next-line import/no-extraneous-dependencies
import { createTSProjectTestContext } from '@fresha/openapi-codegen-test-utils';
import { getAPIName, getRootUrlOrThrow } from '@fresha/openapi-codegen-utils';

import type { Context } from './context';
import type { OpenAPIModel } from '@fresha/openapi-model/build/3.0.3';

export const createTestContext = (openapi: OpenAPIModel, rootDir = '/'): Context => {
  const base = createTSProjectTestContext(openapi, rootDir);

  let apiRootUrl: string;
  try {
    apiRootUrl = getRootUrlOrThrow(openapi);
  } catch (err) {
    apiRootUrl = 'API_URL';
  }

  return {
    ...base,
    apiName: getAPIName(openapi),
    apiRootUrl,
  };
};
