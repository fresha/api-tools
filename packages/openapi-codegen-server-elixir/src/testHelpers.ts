import { Project } from '@fresha/code-morph-ex';
import { createRegistry } from '@fresha/json-api-model';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createTestContext as createBaseTestContext } from '@fresha/openapi-codegen-test-utils';
import { OpenAPIFactory } from '@fresha/openapi-model/build/3.0.3';

import { Generator } from './parts';

import type { Context } from './context';

export const createTestContext = (phoenixApp: string, rootDir = '/'): Context => {
  const openapi = OpenAPIFactory.create();
  const base = createBaseTestContext(openapi, rootDir);

  const project = new Project({
    rootDir,
    phoenixApp,
    useInMemoryFileSystem: true,
  });

  return {
    ...base,
    testObjectFactoryModuleName: `${project.getAppModuleName()}.Factory`,
    openapi,
    project,
    registry: createRegistry(),
  };
};

export const createGenerator = (phoenixApp = 'api_tools_web'): Generator => {
  const context = createTestContext(phoenixApp, '/');
  const generator = new Generator(context);
  return generator;
};
