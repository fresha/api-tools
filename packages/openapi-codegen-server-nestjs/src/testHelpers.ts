import { Nullable } from '@fresha/api-tools-core';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createTSProjectTestContext } from '@fresha/openapi-codegen-test-utils';
import { OpenAPIFactory, SchemaModel } from '@fresha/openapi-model/build/3.0.3';

import { Context } from './context';
import { Generator } from './Generator';

export const createTestContext = (nestApp = 'web', outputPath = '/tmp/'): Context => {
  const openapi = OpenAPIFactory.create();
  const base = createTSProjectTestContext(openapi, outputPath);

  base.project.createSourceFile(
    `${outputPath}/${nestApp}.module.ts`,
    `import { Module } from '@nestjs/common';
    import { AppController } from './app.controller';
    import { AppService } from './app.service';

    @Module({
      imports: [],
      controllers: [AppController],
      providers: [AppService],
    })
    export class AppModule {}`,
  );

  return {
    ...base,
    nestApp,
    addDTO: jest.fn(),
  };
};

export const createGenerator = (nestApp = 'app', rootDir = '/'): Generator => {
  const context = createTestContext(nestApp, rootDir);

  let generator: Generator;
  // eslint-disable-next-line prefer-const
  generator = new Generator({
    ...context,
    addDTO: (name: string, schema: Nullable<SchemaModel>) => generator.addDTO(name, schema),
  });

  return generator;
};
