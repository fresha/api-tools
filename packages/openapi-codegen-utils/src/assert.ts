import nodeAssert from 'assert';

import type { OperationModel } from '@fresha/openapi-model/build/3.0.3';

export function assert(cond: unknown, message: string, operation: OperationModel): asserts cond {
  nodeAssert(
    cond,
    `${message}. Operation (${operation.httpMethod.toUpperCase()} '${operation.parent.pathUrl}')`,
  );
}
