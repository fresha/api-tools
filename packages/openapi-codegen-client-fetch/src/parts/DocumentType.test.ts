import assert from 'assert';

import {
  getOperationDefaultResponseSchemaOrThrow,
  MEDIA_TYPE_JSON_API,
  setResourceSchema,
} from '@fresha/openapi-codegen-utils';
import { OpenAPIFactory, OperationModel, SchemaModel } from '@fresha/openapi-model/build/3.0.3';

import { createActionTestContext } from '../testHelpers';

import { DocumentType } from './DocumentType';

import type { NamedType } from './NamedType';

const createDocumentType = (operation: OperationModel, name: string): DocumentType => {
  const context = createActionTestContext(operation, '/src/index.ts');
  const schema = getOperationDefaultResponseSchemaOrThrow(operation, true);
  assert(schema);
  return new DocumentType(context, name, schema, false, true);
};

let operation: OperationModel;
let namedTypes: Map<string, NamedType>;
let generatedTypes: Set<string>;

beforeEach(() => {
  operation = OpenAPIFactory.create().setPathItem('/hello').addOperation('get');
  namedTypes = new Map<string, NamedType>();
  generatedTypes = new Set<string>();
});

test('generic', () => {
  operation.setDefaultResponse('test').setMediaType(MEDIA_TYPE_JSON_API).setSchema(null);

  const documentType = createDocumentType(operation, 'SimpleResponseDocument');

  documentType.collectData(namedTypes);
  documentType.generateCode(generatedTypes);

  expect(
    documentType.context.project.getSourceFileOrThrow('src/types.ts').getText(),
  ).toMatchSnapshot('src/types.ts');
});

test('jsdocs', () => {
  const responseSchema = operation
    .setDefaultResponse('default response')
    .setMediaType(MEDIA_TYPE_JSON_API)
    .setSchema('object');
  responseSchema.title = 'A simple object schema';
  responseSchema.description =
    'This is a description of the schema.\nThe schema represents a simple object.';
  responseSchema.setProperties({
    jsonapi: { type: 'object', required: true },
    data: { type: 'object', required: true },
  });

  const employee = responseSchema.root.components.setSchema('Employee', 'object');
  employee.setProperties({
    type: { type: 'string', required: true, enum: ['employees'] },
    id: { type: 'string', required: true },
    attributes: { type: 'object', required: true },
  });
  responseSchema.setProperty('data', employee);

  const documentType = createDocumentType(operation, 'SimpleResponseDocument');

  documentType.collectData(namedTypes);
  documentType.generateCode(generatedTypes);

  expect(
    documentType.context.project.getSourceFileOrThrow('src/types.ts').getText(),
  ).toMatchSnapshot('src/types.ts');
});

describe('primary data', () => {
  let responseSchema: SchemaModel;

  beforeEach(() => {
    responseSchema = operation
      .setDefaultResponse('test')
      .setMediaType(MEDIA_TYPE_JSON_API)
      .setSchema('object');
    responseSchema.setProperties({
      jsonapi: { type: 'object', required: true },
      data: { type: 'object', required: true },
    });
  });

  test('single shared resource', () => {
    const employee = responseSchema.root.components.setSchema('Employee', 'object');
    employee.setProperties({
      type: { type: 'string', required: true, enum: ['employees'] },
      id: { type: 'string', required: true },
      attributes: { type: 'object', required: true },
    });
    responseSchema.setProperty('data', employee);

    const documentType = createDocumentType(operation, 'SimpleResponseDocument');

    documentType.collectData(namedTypes);
    documentType.generateCode(generatedTypes);

    expect(
      documentType.context.project.getSourceFileOrThrow('src/types.ts').getText(),
    ).toMatchSnapshot('src/types.ts');
  });

  test('single inline resource', () => {
    const primaryData = responseSchema.getPropertyOrThrow('data');
    primaryData.setProperties({
      type: { type: 'string', required: true, enum: ['employees'] },
      id: { type: 'string', required: true },
      attributes: { type: 'object', required: true },
    });

    const documentType = createDocumentType(operation, 'SimpleResponseDocument');

    documentType.collectData(namedTypes);
    documentType.generateCode(generatedTypes);

    expect(
      documentType.context.project.getSourceFileOrThrow('src/types.ts').getText(),
    ).toMatchSnapshot('src/types.ts');
  });

  test('single resource of possibly multiple types', () => {
    const primaryData = responseSchema.getPropertyOrThrow('data');
    primaryData.type = null;

    const employee = primaryData.addOneOf(null);
    setResourceSchema(employee, 'employees');

    const organization = primaryData.addOneOf(null);
    setResourceSchema(organization, 'organizations');

    const documentType = createDocumentType(operation, 'ResponseDocumentWithUnionTypes');

    documentType.collectData(namedTypes);
    documentType.generateCode(generatedTypes);

    expect(
      documentType.context.project.getSourceFileOrThrow('src/types.ts').getText(),
    ).toMatchSnapshot('src/types.ts');
  });

  test('multiple resources of the same type', () => {
    const employee = responseSchema.root.components.setSchema('Employee', 'object');
    employee.setProperties({
      type: { type: 'string', required: true, enum: ['employees'] },
      id: { type: 'string', required: true },
      attributes: { type: 'object', required: true },
    });

    responseSchema.setProperty('data', 'array').setItems(employee);

    const documentType = createDocumentType(operation, 'SimpleResponseDocument');

    documentType.collectData(namedTypes);
    documentType.generateCode(generatedTypes);

    expect(
      documentType.context.project.getSourceFileOrThrow('src/types.ts').getText(),
    ).toMatchSnapshot('src/types.ts');
  });

  test('skip resources that were generated before', () => {
    const employee = responseSchema.root.components.setSchema('Employee', 'object');
    employee.setProperties({
      type: { type: 'string', required: true, enum: ['employees'] },
      id: { type: 'string', required: true },
      attributes: { type: 'object', required: true },
    });

    responseSchema.setProperty('data', 'array').setItems(employee);

    // pretend that this resource is used in multiple documents, and rendered before
    generatedTypes.add('Employee');

    const documentType = createDocumentType(operation, 'SimpleResponseDocument');

    documentType.collectData(namedTypes);
    documentType.generateCode(generatedTypes);

    expect(
      documentType.context.project.getSourceFileOrThrow('src/types.ts').getText(),
    ).toMatchSnapshot('src/types.ts');
  });
});

describe('included', () => {
  let responseSchema: SchemaModel;

  beforeEach(() => {
    responseSchema = operation
      .setDefaultResponse('test')
      .setMediaType(MEDIA_TYPE_JSON_API)
      .setSchema('object');
    responseSchema.setProperties({
      jsonapi: { type: 'object', required: true },
      data: { type: 'object', required: true },
    });
  });

  test('single type of included resources', () => {
    const employee = responseSchema.root.components.setSchema('Employee', 'object');
    employee.setProperties({
      type: { type: 'string', required: true, enum: ['employees'] },
      id: { type: 'string', required: true },
      attributes: { type: 'object', required: true },
    });

    responseSchema.setProperty('data', employee);
    responseSchema.setProperty('included', 'array').setItems(employee);

    const documentType = createDocumentType(operation, 'DocumentWithIncludedResources');

    documentType.collectData(namedTypes);
    documentType.generateCode(generatedTypes);

    expect(
      documentType.context.project.getSourceFileOrThrow('src/types.ts').getText(),
    ).toMatchSnapshot('src/types.ts');
  });

  test('includes resources of multiple types', () => {
    const employee = responseSchema.root.components.setSchema('Employee', 'object');
    employee.setProperties({
      type: { type: 'string', required: true, enum: ['employees'] },
      id: { type: 'string', required: true },
      attributes: { type: 'object', required: true },
    });

    const organization = responseSchema.root.components.setSchema('Organization', 'object');
    organization.setProperties({
      type: { type: 'string', required: true, enum: ['organizations'] },
      id: { type: 'string', required: true },
      attributes: { type: 'object', required: true },
    });

    responseSchema.setProperty('data', employee);

    const includedSchema = responseSchema.setProperty('included', 'array');
    const includedItemSchema = includedSchema.setItems(null);
    includedItemSchema.addOneOf(organization);
    includedItemSchema.addOneOf(employee);

    const documentType = createDocumentType(operation, 'DocumentWithIncludedResources');

    documentType.collectData(namedTypes);
    documentType.generateCode(generatedTypes);

    expect(
      documentType.context.project.getSourceFileOrThrow('src/types.ts').getText(),
    ).toMatchSnapshot('src/types.ts');
  });
});
