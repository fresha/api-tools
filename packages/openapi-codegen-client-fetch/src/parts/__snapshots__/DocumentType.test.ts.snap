// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generic: src/types.ts 1`] = `
"import type { JSONAPIDataDocument } from "@fresha/api-tools-core";


/**
 * This is a response resource
 *
 */
export type SimpleResponseDocument = JSONAPIDataDocument;
"
`;

exports[`included includes resources of multiple types: src/types.ts 1`] = `
"import type { JSONAPIServerResource, JSONAPIDataDocument } from "@fresha/api-tools-core";


/**
 * Employee
 *
 */
export type Employee = JSONAPIServerResource<'employees'>;

/**
 * Organization
 *
 */
export type Organization = JSONAPIServerResource<'organizations'>;

/**
 * This is a response resource
 *
 * Primary data resources:
 *  - Employee
 *
 * Included resources:
 *  - Organization
 *  - Employee
 *
 */
export type DocumentWithIncludedResources = JSONAPIDataDocument<Employee, Organization | Employee>;
"
`;

exports[`included single type of included resources: src/types.ts 1`] = `
"import type { JSONAPIServerResource, JSONAPIDataDocument } from "@fresha/api-tools-core";


/**
 * Employee
 *
 */
export type Employee = JSONAPIServerResource<'employees'>;

/**
 * This is a response resource
 *
 * Primary data resources:
 *  - Employee
 *
 * Included resources:
 *  - Employee
 *
 */
export type DocumentWithIncludedResources = JSONAPIDataDocument<Employee, Employee>;
"
`;

exports[`jsdocs: src/types.ts 1`] = `
"import type { JSONAPIServerResource, JSONAPIDataDocument } from "@fresha/api-tools-core";


/**
 * Employee
 *
 */
export type Employee = JSONAPIServerResource<'employees'>;

/**
 * A simple object schema
 *
 * This is a description of the schema.
 * The schema represents a simple object.
 *
 * This is a response resource
 *
 * Primary data resources:
 *  - Employee
 *
 */
export type SimpleResponseDocument = JSONAPIDataDocument<Employee>;
"
`;

exports[`primary data multiple resources of the same type: src/types.ts 1`] = `
"import type { JSONAPIServerResource, JSONAPIDataDocument } from "@fresha/api-tools-core";


/**
 * Employee
 *
 */
export type Employee = JSONAPIServerResource<'employees'>;

/**
 * This is a response resource
 *
 * Primary data resources:
 *  - Employee
 *
 */
export type SimpleResponseDocument = JSONAPIDataDocument<(Employee)[]>;
"
`;

exports[`primary data single inline resource: src/types.ts 1`] = `
"import type { JSONAPIServerResource, JSONAPIDataDocument } from "@fresha/api-tools-core";


export type Unknown1 = JSONAPIServerResource<'employees'>;

/**
 * This is a response resource
 *
 * Primary data resources:
 *  - Unknown1
 *
 */
export type SimpleResponseDocument = JSONAPIDataDocument<Unknown1>;
"
`;

exports[`primary data single resource of possibly multiple types: src/types.ts 1`] = `
"import type { JSONAPIServerResource, JSONAPIDataDocument } from "@fresha/api-tools-core";


/**
 * EmployeeResource
 *
 */
export type EmployeeResource = JSONAPIServerResource<'employees'>;

/**
 * OrganizationResource
 *
 */
export type OrganizationResource = JSONAPIServerResource<'organizations'>;

/**
 * This is a response resource
 *
 * Primary data resources:
 *  - EmployeeResource
 *  - OrganizationResource
 *
 */
export type ResponseDocumentWithUnionTypes = JSONAPIDataDocument<EmployeeResource | OrganizationResource>;
"
`;

exports[`primary data single shared resource: src/types.ts 1`] = `
"import type { JSONAPIServerResource, JSONAPIDataDocument } from "@fresha/api-tools-core";


/**
 * Employee
 *
 */
export type Employee = JSONAPIServerResource<'employees'>;

/**
 * This is a response resource
 *
 * Primary data resources:
 *  - Employee
 *
 */
export type SimpleResponseDocument = JSONAPIDataDocument<Employee>;
"
`;

exports[`primary data skip resources that were generated before: src/types.ts 1`] = `
"import type { JSONAPIDataDocument } from "@fresha/api-tools-core";


/**
 * This is a response resource
 *
 * Primary data resources:
 *  - Employee
 *
 */
export type SimpleResponseDocument = JSONAPIDataDocument<(Employee)[]>;
"
`;
